docs/new_protocol.asciidoc

```diff
@@ -0,0 +1,463 @@
+is detected in a TCP stream. The protocol module can use these callbacks to take
+decisions about what to do with partial data received. For example, for the
+HTTP/1.0 protocol, the end of connection is used to know when the message is
+finished.
+
+
+==== The TCP Parse function
+
+For TCP protocols, the `Parse()` function is the heart of the module. As
+mentioned before, it is called for every TCP packet containing data on the
+configured ports.
+
+It is important to understand that because TCP is a stream
+based protocol, the packets boundaries don't necessarily match the application
+layer message boundaries. For example, a packet can contain only a part of the
+message, it can contain a complete message or it can contain multiple messages.
+
+If you see a packet in the middle of the stream, you have no guaranties that its
+first byte is the beginning of a message. However, if the packet is the first
+seen in a given TCP stream, that you can assume is the beginning of the message.
+
+The Parse function needs to deal with these facts which generally means that it
+needs to keep state across multiple packets.
+
+Let's have a look again at its signature:
+
+[source,go]
+----------------------------------------------------------------------
+func Parse(pkt *protos.Packet, tcptuple *common.TcpTuple, dir uint8,
+	private protos.ProtocolData) protos.ProtocolData
+----------------------------------------------------------------------
+
+We've already talked about the first parameter which contains the packet data.
+The rest of the parameters and the return value are used for maintaining state
+inside the TCP stream.
+
+The `tcptuple` is an unique identifier for the TCP stream from which the packet
+is part of. You can use the `tcptuple.Hashable()` functions to get a value that
+you can store in a map. The `dir` flag gives you the direction in which the
+packet is flowing inside the TCP stream. The two possible values are
+`TcpDirectionOriginal` if the packet goes in the same direction as the first
+packet that we saw from that stream and `TcpDirectionReverse` if the packet goes
+the other direction.
+
+The `private` parameter can be used by the module to store in the TCP stream
+whatever state it needs. The module would typically cast this at runtime to a
+type of its choice, modify it as needed and then return the modified value.
+Next time the TCP layer calls the `Parse()` or the others functions from the
+`TcpProtocolPlugin` interface, it will call it with the modified private value.
+
+Here is an example handling of the private data as it's done by the MySQL
+module:
+
+[source,go]
+----------------------------------------------------------------------
+	priv := mysqlPrivateData{}
+	if private != nil {
+		var ok bool
+		priv, ok = private.(mysqlPrivateData)
+		if !ok {
+			priv = mysqlPrivateData{}
+		}
+	}
+
+	[ ... ]
+
+	return priv
+----------------------------------------------------------------------
```

Comment:
Something I tripped on is the private element being copied all the time. Just return nil to drop current state is easy though. => if private data element becomes large consider using pointers. Plus be aware of copying semantics in Gap/Fin handlers